frame/composable-traits/src/defi.rs

```diff
@@ -70,15 +73,21 @@ pub struct CurrencyPair<AssetId> {
 }
 
 impl<AssetId: PartialEq> CurrencyPair<AssetId> {
-	pub fn is_valid(&self) -> bool {
-		self.base != self.quote
+	pub fn new(base: AssetId, quote: AssetId) -> Self {
+		Self { base, quote }
+	}
+
+	pub fn as_array(&self) -> &[AssetId; 2] {
+		// adding lifetimes explicitly considered by rust as not needed, so lifetime is proper for
+		// return array
+		unsafe { sp_std::mem::transmute(self) }
 	}
 }
 
 /// type parameters for traits in pure defi area
 pub trait DeFiEngine {
 	/// The asset ID type
-	type AssetId: AssetIdLike;
+	type MayBeAssetId: AssetIdLike;
```

Comment:
will this naming convention apply to all other associated types like `Balance` ?

Reply:
It is only about id. Really this type is never exactly asseit id by always may be id. Tokens type is about using that MayBeId (I assume these must error if ID does not exist)